### PR TITLE
Zone file import - support missing params.

### PR DESCRIPTION
**Pull request by hhellyer**
This PR adds support for missing parameters when importing zone files:
- networks
- views
- name

As both zone creation with or without a zone file to prepopulate it use the same call, just with or without the zonefile specificed, the "name" field also added to the api to create a zone and "views" is added to the list of fields passed through on zone creation.

### CODE REVIEW

**Comment by hhellyer** [comment]
Since this will need an update to the changelog before releasing I've added my proposed changelog update below. (It is a bit wordy because of the need to differentiate between behaviour changes when creating a zone and new things when creating a zone with a zonefile.)
```
## 0.20.1 (June ??th, 2024)

ENHANCEMENTS:
* Adds support for specifying a list of views when creating zones with or without a provided zone file.
* Adds support for specifying a zone name other than the FQDN when creating zones with or without a provided zone file.
* A specified list of networks for a zone was only applied to zone creation when a zone file was not provided. If a list of networks is specified it will now be applied when creating zones by uploading a zone file as well.
```

File: ns1/__init__.py [hunk 1]
```diff
@@ -371,7 +377,8 @@ def createZone(
         If zoneFile is specified, upload the specific zone definition file
         to populate the zone with.
 
-        :param str zone: zone name, like 'example.com'
+        :param str zone: zone FQDN, like 'example.com'
+        :param str zone: zone name override, name will be zone FQDN if omitted
```
Comment (shane-ns1) [inline ns1/__init__.py:381]:
```suggestion
        :param str name: zone name override, name will be zone FQDN if omitted
```

It is possible in Python to declare keyword-only arguments:

https://peps.python.org/pep-3102/

But we don't have that now, so order matters. 

If you want to reorder the parameters like this you'll need to make a major release, as it can break the API. Alternately you can put it at the end. :smile:

File: ns1/rest/zones.py [hunk 2]
```diff
@@ -21,32 +21,59 @@ class Zones(resource.BaseResource):
...
+
     def _buildBody(self, zone, **kwargs):
         body = {}
         body["zone"] = zone
         self._buildStdBody(body, kwargs)
         return body
 
     def import_file(
-        self, zone, zoneFile, callback=None, errback=None, **kwargs
+        self, zone, zoneFile, name=None, callback=None, errback=None, **kwargs
```
Comment (shane-ns1) [inline ns1/rest/zones.py:40]: Similar to the change in `__init__.py`, this reorders the parameters and is a breaking change.

File: ns1/rest/zones.py [hunk 3]
```diff
@@ -21,32 +21,59 @@ class Zones(resource.BaseResource):
...
 
-    def create(self, zone, callback=None, errback=None, **kwargs):
+    # Extra import args are specified as query parameters not fields in a JSON object.
+    def _buildImportParams(self, fields, name=None):
+        params = {}
+        # Arrays of values should be passed as multiple instances of the same
+        # parameter but the zonefile API expects parameters containing comma
+        # seperated values.
+        if fields.get("networks") is not None:
+            networksStrs = map(str, fields["networks"])
```
Comment (shane-ns1) [inline ns1/rest/zones.py:60]: I guess this is because people may pass networks as numbers?

File: ns1/rest/zones.py [hunk 4]
```diff
@@ -21,32 +21,59 @@ class Zones(resource.BaseResource):
...
+            networksParam = ",".join(networksStrs)
+            params["networks"] = networksParam
+        if fields.get("views") is not None:
```
Comment (shane-ns1) [inline ns1/rest/zones.py:63]: But we don't expect people to pass views as numbers?

File: ns1/rest/zones.py [hunk 5]
```diff
@@ -21,32 +21,59 @@ class Zones(resource.BaseResource):
...
+            viewsParam = ",".join(fields["views"])
+            params["views"] = viewsParam
+        if name is not None:
+            params["name"] = name
+        return params
+
+    def create(self, zone, name=None, callback=None, errback=None, **kwargs):
```
Comment (shane-ns1) [inline ns1/rest/zones.py:70]: Here also, changing the order of parameters may break things for some users.

**Review by shane-ns1 — CHANGES_REQUESTED** [review]
There's a typo (`zone` instead of `name` in a docstring), but more important is that this changes order of parameters passed to the API.

File: ns1/rest/zones.py (hunk 3 above)
Comment (cbertinato) [inline ns1/rest/zones.py:60]: Snake case in Python.

**Review by cbertinato — CHANGES_REQUESTED** [review]

File: ns1/rest/zones.py (hunk 3 above)
Comment (hhellyer) [reply ns1/rest/zones.py:60]: Our API takes a list of network numbers.
Unfortunately it takes them in a single query parameter of the form `networks=1,2,3` which is *not* how you should pass an array via parameters, as I've found out while working on the other side of our import API, and not how the request library will pass an array. So we need to convert the array to a string.
If allowed to the libs will pass an array using multiple params of the same name:
`networks=1&networks=2&networks=3`
but our API won't accept that.

File: ns1/rest/zones.py (hunk 4 above)
Comment (hhellyer) [reply ns1/rest/zones.py:63]: No, views are passed by name.
🤷

File: ns1/rest/zones.py [hunk 6]
```diff
@@ -21,32 +21,59 @@ class Zones(resource.BaseResource):
...
         body = self._buildBody(zone, **kwargs)
+        if name is None:
+            name = zone
         return self._make_request(
             "PUT",
-            "%s/%s" % (self.ROOT, zone),
+            "%s/%s" % (self.ROOT, name),
```
[the remark this answers is not in the document]
Comment (hhellyer) [reply ns1/rest/zones.py]: Done, on this function, `create()`, and on `import()`. I've left the others in this file as those were the two I altered and (more importantly) wrote tests for. There aren't specific tests for the other functions in this file.

File: ns1/rest/zones.py [hunk 7]
```diff
@@ -34,19 +40,40 @@ def import_file(
...
 
-    def create(self, zone, callback=None, errback=None, **kwargs):
+    # Extra import args are specified as query parameters not fields in a JSON object.
+    def _buildImportParams(self, fields):
+        params = {}
+        # Arrays of values should be passed as multiple instances of the same
+        # parameter but the zonefile API expects parameters containing comma
+        # seperated values.
+        if fields.get("networks") is not None:
+            networks_strs = map(str, fields["networks"])
```
Comment (shane-ns1) [inline ns1/rest/zones.py:60]: I think this is more Pythonic:
```suggestion
            networks_strs = [str(network) for network in fields["networks"]]
```

File: tests/unit/test_zone.py [hunk 8]
```diff
@@ -64,6 +65,75 @@ def test_rest_zone_version_list(zones_config, zone, url):
...
+    networksStrs = None
+    if networks is not None:
+        networksStrs = map(str, networks)
+        params["networks"] = ",".join(networksStrs)
+    if views is not None:
+        params["views"] = ",".join(views)
+    if name is not None:
+        params["name"] = name
+
+    zoneFilePath = "{}/../../examples/importzone.db".format(
```
[the remark this answers is not in the document]
Comment (cbertinato) [reply tests/unit/test_zone.py:154]: Remaining comment, but also a nitpick. So change it if you'd like, but I won't block on it.

**Review by cbertinato — APPROVED** [review]

File: ns1/__init__.py (hunk 1 above)
Comment (shane-ns1) [reply ns1/__init__.py]: We're < 1.0 so we don't have to do a major release for such things (I think). But we will be breaking stuff!

**Review by shane-ns1 — APPROVED** [review]
